File: src/java/pulse/cdm/properties/SEScalarInversePressure.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

package pulse.cdm.properties;

import pulse.cdm.bind.Properties.ScalarInversePressureData;
import pulse.cdm.properties.CommonUnits.InversePressureUnit;

/**
 * @author abray
 * InversePressure Scalar Class, will enforce that units are proper to InversePressures
 */

public class SEScalarInversePressure extends SEScalar
{
  public SEScalarInversePressure()
  {
    super();
  }
  
  /**
   * @param value
   * @param unit - specific any valid standard unit abbreviation
   * http://www.bipm.org/en/si/ for this quantity type
   */
  public SEScalarInversePressure(double value, String unit)
  {
    this();
    this.setValue(value, unit);
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public SEScalarInversePressure(double value, InversePressureUnit unit)
  {
    this();
    this.setValue(value,unit);
  }
  
  public static void load(ScalarInversePressureData src, SEScalarInversePressure dst)
  {
    SEScalar.load(src.getScalarInversePressure(),dst);
  }
  public static ScalarInversePressureData unload(SEScalarInversePressure src)
  {
    if(!src.isValid())
      return null;
    ScalarInversePressureData.Builder dst = ScalarInversePressureData.newBuilder();
    unload(src,dst);
    return dst.build();
  }
  protected static void unload(SEScalarInversePressure src, ScalarInversePressureData.Builder dst)
  {
    SEScalar.unload(src,dst.getScalarInversePressureBuilder());
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public void setValue(double value, InversePressureUnit unit)
  {
    this.setValue(value,unit.toString());
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public double getValue(InversePressureUnit unit)
  {
    return this.getValue(unit.toString());
  }

  @Override
  public boolean validUnit(String unit)
  {
    if(InversePressureUnit.validUnit(unit))
      return true;
    return false;
  }
}

File: src/csharp/pulse/cdm/properties/SEScalarInversePressure.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

namespace Pulse.CDM
{
  public class InversePressureUnit : Unit
  {
    private InversePressureUnit(string v) : base(v) { }

    public static InversePressureUnit Inverse_Pa = new InversePressureUnit("1/Pa");
    public static InversePressureUnit Inverse_mmHg = new InversePressureUnit("1/mmHg");
    public static InversePressureUnit Inverse_cmH2O = new InversePressureUnit("1/cmH2O");
    public static InversePressureUnit Inverse_atm = new InversePressureUnit("1/atm");

    public static InversePressureUnit FromString(string u)
    {
      if (u == InversePressureUnit.Inverse_Pa.ToString())
        return InversePressureUnit.Inverse_Pa;
      if (u == InversePressureUnit.Inverse_mmHg.ToString())
        return InversePressureUnit.Inverse_mmHg;
      if (u == InversePressureUnit.Inverse_cmH2O.ToString())
        return InversePressureUnit.Inverse_cmH2O;
      if (u == InversePressureUnit.Inverse_atm.ToString())
        return InversePressureUnit.Inverse_atm;

      throw new System.ArgumentException(u + " is not a know inverse pressure unit");
    }
  }

  public class SEScalarInversePressure : SEScalar
  {
    protected InversePressureUnit unit;

    public SEScalarInversePressure() : base()
    {

    }

    public SEScalarInversePressure(SEScalarInversePressure from)
    {
      Set(from);
    }

    public SEScalarInversePressure(double value, InversePressureUnit unit)
    {
      SetValue(value, unit);
    }

    public void Set(SEScalarInversePressure from)
    {
      if (from == null)
      {
        Invalidate();
        return;
      }
      SetValue(from.value, from.unit);
    }

    public new void SetValue(double value)
    {
      throw new System.NotSupportedException("You must provide a unit");
    }

    public void SetValue(double value, InversePressureUnit unit)
    {
      this.value = value;
      this.unit = unit;
    }

    public double GetValue(InversePressureUnit unit)
    {
      return UnitConverter.Convert(this.value, this.unit, unit);
    }

    public InversePressureUnit GetUnit()
    {
      return unit;
    }

    public new string ToString()
    {
      return this.value + "(" + this.unit.ToString() + ")";
    }
  }
}
